docs/proposals/20210524-etcdadm-bootstrap-provider.md

```diff
@@ -0,0 +1,564 @@
+
+###### Upgrade or member replacement use case
+
+- The EtcdCluster.Spec.Replicas field will be immutable. This is to protect the etcd cluster from losing quorum by unnecessarily adding/removing healthy members.
+- An etcd member will need to be replaced by the etcd controller in two cases:
+  - Member is found unhealthy
+  - During an upgrade
+- While replacing an unhealthy etcd member, it is important to first remove the member and only then replace it by adding a new one. This ensures the cluster stays stable. [This Etcd doc](https://etcd.io/docs/v3.4/faq/#should-i-add-a-member-before-removing-an-unhealthy-member) explains why removing an unhealthy member is important in depth. Etcd's latest versions allow adding new members as learners so it doesn't affect the quorum size, but that feature is in beta. Also, adding a learner is a two-step process, where you first add a member as learner and then promote it, and etcdadm currently doesn't support it. So we can include this in a later version once the learner feature becomes GA and etcdadm incorporates the required changes.
+- We can reuse this same "scale down" function when removing a healthy member during an etcd upgrade. So even upgrade will follow the same pattern of removing the member first and then adding a new one with the latest spec.
+- Once the upgrade/machine replacement is completed, the etcd controller will again update the EtcdCluster.Status.Endpoint field. The control plane controller will rollout new Machines that use the updated etcd endpoints for the apiserver.
+
+###### Periodic etcd member healthchecks
+- The controller should do healthchecks of the etcd members in a loop at a certain predetermined interval (10s).
+- When a member fails healthcheck for a certain amount of time (1m), replace that member with a new member
+
+**Changes needed in docker machine controller**
+- Each machine infrastructure provider sets a providerID on every InfraMachine.
+- Infrastructure providers like CAPA and CAPV set this provider ID based on the instance metadata.
+- CAPD on the other hand first calls `kubectl patch` to add the provider ID on the node. And only then it sets the provider ID value on the DockerMachine resource. But the problem is that in case of an etcd only cluster, the machine is not registered as a Kubernetes node. Cluster provisioning does not progress until this provider ID is set on the Machine. As a solution, CAPD will check if the bootstrap provider is etcdadm, and skip the kubectl patch process and set the provider ID on DockerMachine directly. These changes are required in the [docker machine controller](https://github.com/mrajashree/cluster-api/pull/2/files#diff-1923dd8291a9406e3c144763f526bd9797a2449a030f5768178b8d06d13c795bR307) for CAPD.
+
+**Future work**
+
+##### Static Etcd endpoints
```

Comment:
Great callout. Does this mean the first iteration would be a set of IPs that may change? Does kube-apiserver need restart?

Reply:
that's right